vmm/src/main.rs

```diff
@@ -201,22 +211,85 @@ fn main() -> ExitCode {
     let mut vcpu_regs = vcpu_fd.get_regs().unwrap();
     vcpu_regs.rip = start_address;
     let initial_stack = unsafe {
-        Box::<[u8; 2 * 0x400 * 0x400], &BuddyAllocator>::new_uninit_in(&allocator).assume_init()
+        Box::<[u8; 2 * 0x400 * 0x400], &BuddyAllocator>::new_uninit_in(allocator).assume_init()
     };
     let stack_start = allocator.to_offset(&*initial_stack);
     Box::leak(initial_stack);
 
+    let ring_buffer_in: RefCnt<RingBuffer> = RingBuffer::new_in(1024, allocator).into();
+    let ring_buffer_data_in: RefCnt<RingBufferRawData> =
+        Box::new_in(ring_buffer_in.into_raw_parts(allocator), allocator).into();
+
+    let ring_buffer_out: RefCnt<RingBuffer> = RingBuffer::new_in(1024, allocator).into();
+    let ring_buffer_data_out: RefCnt<RingBufferRawData> =
+        Box::new_in(ring_buffer_out.into_raw_parts(allocator), allocator).into();
+
+    let client = Arc::new(RwLock::new(Client::new(ring_buffer_in, ring_buffer_out)));
+
     let raw = allocator.clone().into_raw_parts();
     let pa2ka = |p: usize| (p | 0xFFFF800000000000);
     vcpu_regs.rsp = pa2ka((stack_start + (2 * 0x400 * 0x400)) as usize) as u64;
     vcpu_regs.rdi = raw.inner_offset as u64;
     vcpu_regs.rsi = raw.inner_size as u64;
     vcpu_regs.rdx = raw.refcnt_offset as u64;
     vcpu_regs.rcx = raw.refcnt_size as u64;
+    vcpu_regs.r8 = allocator.to_offset(RefCnt::into_raw(ring_buffer_data_out.clone())) as u64;
+    vcpu_regs.r9 = allocator.to_offset(RefCnt::into_raw(ring_buffer_data_in.clone())) as u64;
     vcpu_regs.rflags = 2;
     vcpu_fd.set_regs(&vcpu_regs).unwrap();
 
     let mut record: usize = 0;
+    let client_handle = thread::spawn(move || {
+        let mut addelf =
+            unsafe { Arc::new_uninit_slice_in(ADD_ELF.len(), allocator).assume_init() };
+        Arc::get_mut(&mut addelf).unwrap().copy_from_slice(ADD_ELF);
+
+        let x = 1 as u64;
+        let y = 1 as u64;
+        let mut xblob = unsafe { Arc::new_uninit_slice_in(8, allocator).assume_init() };
+        let mut yblob = unsafe { Arc::new_uninit_slice_in(8, allocator).assume_init() };
+        Arc::get_mut(&mut xblob)
+            .unwrap()
+            .copy_from_slice(&x.to_ne_bytes());
+        Arc::get_mut(&mut yblob)
+            .unwrap()
+            .copy_from_slice(&y.to_ne_bytes());
+
+        client.write().unwrap().create_blob(addelf);
+        client.write().unwrap().create_blob(xblob);
+        client.write().unwrap().create_blob(yblob);
+        client.write().unwrap().send_all();
```

Comment:
All this repeated locking and unwrapping is messy. Does the main thread even need to access this client? Why not just move it into the thread directly and avoid the lock?